caikit/runtime/service_generation/data_stream_source.py

```diff
@@ -255,53 +202,354 @@ def _get_resolved_source_path(input_path: str) -> str:
+        self._plugins = None
+
+    def get_plugins(
+        self, plugins_config: Optional[aconfig.Config] = None
+    ) -> List[DataStreamSourcePlugin]:
+        """Builds the set of plugins to use for a data stream source of type element_type"""
+        if self._plugins is None:
+            self._plugins = []
+            if plugins_config is None:
+                plugins_config = get_config().data_streams.source_plugins
+            for name, cfg in plugins_config.items():
+                self._plugins.append(self.construct(cfg, name))
+
+            # Make sure field numbers are unique
+            field_numbers = [plugin.get_field_number() for plugin in self._plugins]
+            duplicates_field_numbers = [
+                plugin.name
+                for plugin in self._plugins
+                if field_numbers.count(plugin.get_field_number()) > 1
+            ]
+            error.value_check(
+                "<COR69189361E>",
+                not duplicates_field_numbers,
+                "Duplicate plugin field numbers found for plugins: {}",
+                duplicates_field_numbers,
+            )
+        return self._plugins
+
+
+# Single default instance
+PluginFactory = DataStreamPluginFactory("DataStreamSource")
+PluginFactory.register(JsonDataStreamSourcePlugin)
+PluginFactory.register(FileDataStreamSourcePlugin)
+PluginFactory.register(ListOfFilesDataStreamSourcePlugin)
+PluginFactory.register(DirectoryDataStreamSourcePlugin)
+PluginFactory.register(S3FilesDataStreamSourcePlugin)
+
+
+## DataStreamSourceBase ########################################################
+
+
+class DataStreamSourceBase(DataStream):
+    """This base class acts as a sentinel so that dynamically generated data
+    stream source classes can be identified programmatically.
+    """
+
+    def __init__(self):
+        super().__init__(self._generator)
+
+    def _generator(self):
+        stream = self.to_data_stream()
+        return stream.generator_func(*stream.generator_args, **stream.generator_kwargs)
+
+    def __getstate__(self) -> bytes:
+        """A DataStreamSource is pickled by serializing its source
+        representation. This is particularly useful when sharing data streams
+        across subprocesses to run training in an isolated process.
+        """
+        return self.to_binary_buffer()
+
+    def __setstate__(self, pickle_bytes: bytes):
+        """Unpickling a DataStreamSource basically involves unpacking the
+        serialized source representation. The catch is that the oneof is
+        represented strangely in __dict__, so we need to explicitly set all
+        oneof members.
+        """
+        new_inst = self.__class__.from_binary_buffer(pickle_bytes)
+        setattr(self, new_inst.which_oneof("data_stream"), new_inst.data_stream)
+        self.generator_func = self._generator
+        self.generator_args = tuple()
+        self.generator_kwargs = {}
+
+    @property
+    def name_to_plugin_map(self):
```

Comment:
I think we might be able to do this the "official" way: https://stackoverflow.com/a/64647117